### PR TITLE
Add support for the JCIP and Javax @GuardedBy annotations

### PR DESCRIPTION
**Pull request by jthaine**
Note that the JCIPAnnotations.java test cases are now greatly simplified because JCIP and Javax @GuardedBy are not allowed on receivers.

### CODE REVIEW

File: checker/manual/lock-checker.tex [hunk 1]
```diff
@@ -872,6 +872,54 @@ \subsection{Discussion of \<@Holding>\label{lock-checker-holding}}
...
+elsewhere.
+
+If your code is already annotated with a different lock
+annotation, you can reuse that effort.  The Checker Framework comes with
+cleanroom re-implementations of annotations from other tools.  It treats
+them exactly as if you had written the corresponding annotation from the
+Lock Checker, as described in Figure~\ref{fig-lock-refactoring}.
+
+
+\begin{figure}
```
Comment (smillst) [inline checker/manual/lock-checker.tex]: This table looks weird.

File: checker/manual/lock-checker.tex [hunk 2]
```diff
@@ -872,6 +872,54 @@ \subsection{Discussion of \<@Holding>\label{lock-checker-holding}}
...
+\end{center}
+%BEGIN LATEX
+\vspace{-1.5\baselineskip}
+%END LATEX
+\caption{Correspondence between other lock annotations and the
+  Checker Framework's annotations.}
+\label{fig-lock-refactoring}
+\end{figure}
+
+Alternately, the Checker Framework can process those other annotations (as
```
Comment (smillst) [inline checker/manual/lock-checker.tex]: This paragraph doesn't make sense and seems to be a repeat of the one above it.

File: checker/src/javax/annotation/concurrent/GuardedBy.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import org.checkerframework.framework.qual.PreconditionAnnotation;
+
+/**
+ * GuardedBy
+ *
+ * The field or method to which this annotation is applied can only be accessed
+ * when holding a particular lock, which may be a built-in (synchronization)
+ * lock, or may be an explicit java.util.concurrent.Lock.
+ *
+ * The argument determines which locks guard the annotated field or method: this :
```
Comment (smillst) [inline checker/src/javax/annotation/concurrent/GuardedBy.java]: This isn't formatted well.  Can you copy the comment from the original again?

File: checker/manual/lock-checker.tex [hunk 4]
```diff
@@ -872,6 +872,54 @@ \subsection{Discussion of \<@Holding>\label{lock-checker-holding}}
...
 
 
+\section{Other lock annotations\label{lock-other-annotations}}
+
+The Checker Framework's lock annotations are similar to annotations used
+elsewhere.
+
+If your code is already annotated with a different lock
+annotation, you can reuse that effort.  The Checker Framework comes with
+cleanroom re-implementations of annotations from other tools.  It treats
```
Comment (smillst) [inline checker/manual/lock-checker.tex]: The JCIP GuardedBy isn't a cleanroom re-implementation.  It's a copy.

File: checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java [hunk 5]
```diff
@@ -511,4 +515,49 @@ protected TreeAnnotator createTreeAnnotator() {
...
+        if (tree.getKind() == Tree.Kind.VARIABLE) {
+
+            // If one of these declaration annotations is present on a variable declaration:
+            // @net.jcip.annotations.GuardedBy
+            // @javax.annotation.concurrent.GuardedBy
+            // insert the corresponding
+            // @org.checkerframework.checker.lock.qual.GuardedBy
+            // type qualifier.
+
+            List<AnnotationMirror> annos = InternalUtils
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java]: I would use AnnotatedTypeFactory#getDeclAnnotation(Element, java.lang.Class<? extends Annotation>) instead.

File: checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java [hunk 6]
```diff
@@ -511,4 +515,49 @@ protected TreeAnnotator createTreeAnnotator() {
                new ImplicitsTreeAnnotator(this)
         );
     }
+
+    @Override
+    public void annotateImplicit(Tree tree, AnnotatedTypeMirror type, boolean useFlow) {
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java]: You'll need to do the same thing for annotateImplicit(Element, ...)

File: checker/manual/lock-checker.tex (hunk 4 above)
Comment (mernst) [inline checker/manual/lock-checker.tex]: If so, then the current versions need to be deleted and re-implemented, for copyright/licensing issues.

File: checker/src/javax/annotation/concurrent/GuardedBy.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
+package javax.annotation.concurrent;
+
+
+/*
+ * Copyright (c) 2005 Brian Goetz
+ * Released under the Creative Commons Attribution License
+ *   (http://creativecommons.org/licenses/by/2.5)
+ * Official home: http://www.jcip.net
```
Comment (mernst) [inline checker/src/javax/annotation/concurrent/GuardedBy.java]: Please merge upstream into this pull request, which will update this file.

File: checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java [hunk 8]
```diff
@@ -511,4 +515,66 @@ protected TreeAnnotator createTreeAnnotator() {
...
+    /**
+     * Given a field declaration with a {@code @net.jcip.annotations.GuardedBy} or
+     * {@code javax.annotation.concurrent.GuardedBy} annotation and an AnnotatedTypeMirror
+     * for that field, inserts the corresponding {@code @org.checkerframework.checker.lock.qual.GuardedBy}
+     * type qualifier into that AnnotatedTypeMirror.
+     *
+     * @param element the Element for a field declaration (this method does nothing if the Element is not for a field declaration)
+     * @param atm the AnnotatedTypeMirror for the field - the {@code @GuardedBy} type qualifier will be inserted here
+     */
+    private void insertLockCheckerGuardedByIfJcipOrJavaxGuardedByIsPresent(Element element, AnnotatedTypeMirror atm) {
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java]: This is a rather lengthy method name.  What about `translateJcipAndJavacAnnotions` instead?  (Though, if you like the longer one better, it's ok)

File: checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java [hunk 9]
```diff
@@ -511,4 +515,66 @@ protected TreeAnnotator createTreeAnnotator() {
...
+        super.annotateImplicit(tree, type, useFlow);
+    }
+
+    /**
+     * Given a field declaration with a {@code @net.jcip.annotations.GuardedBy} or
+     * {@code javax.annotation.concurrent.GuardedBy} annotation and an AnnotatedTypeMirror
+     * for that field, inserts the corresponding {@code @org.checkerframework.checker.lock.qual.GuardedBy}
+     * type qualifier into that AnnotatedTypeMirror.
+     *
+     * @param element the Element for a field declaration (this method does nothing if the Element is not for a field declaration)
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java]: The parenthetical makes  `the Element for a field declaration` incorrect.  It's really any element.

File: checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java [hunk 10]
```diff
@@ -511,4 +515,66 @@ protected TreeAnnotator createTreeAnnotator() {
...
+     * @param atm the AnnotatedTypeMirror for the field - the {@code @GuardedBy} type qualifier will be inserted here
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java]: should be for `element`.

**Comment by smillst** [comment]
The code changes look good.

**Comment by smillst** [comment]
@mernst Could you please review/edit the manual changes?

**Comment by mernst** [comment]
I already did, and made some changes to the pull request too.

**Pull request merged**